### PR TITLE
Overlapped (#1)

### LINKED ISSUE #80595: Overlapping Optimizer.step() with DDP backward
### 🚀 The feature, motivation and pitch

DDP `all_reduce` parameters' gradients in `buckets`, which means that some parameters would get their finial gradients earlier than others. This exposes an opportunity for `Optimizer` to optimize part of parameters before `backward` finished. So that overlapping part of `Optimizer.step` and DDP `backward` is possible and may led to higher parallelism or even hide some `all_reduce` cost.

### Alternatives

This re-scheduling may be achieved by compilation optimizations like `LazyTensorCore` and `XLA`.

### Additional context

截屏2022-06-30 下午3 58 35

As the timeline above, if we can start `Optimizer.step` working on those parameters already has finial gradients, we may save almost all `Optimizer.step`'s cost.

### PR DESCRIPTION
**Pull request by li-yi-dong**
* Overlapped optimizer

Co-authored-by: liyidong.lyd <liyidong.lyd@alibaba-inc.com>

Fixes #80595

### CODE REVIEW

**Comment by li-yi-dong** [comment]
Here is a timeline profiled on a ResNet50, training on 2 nodes:
Before modification:
image
After:
image

During testing on ResNet, I find out some drawbacks for overlapping `Optim.step` with `backward`:
1. Python frontend for launching `OverlappedOptimizer`'s kernel is more time consuming than original `Optimizer`
2. When enable `GradScaler` together with `OverlappedOptimizer`, there is a significant performance degradation since `_amp_foreach_non_finite_check_and_unscale_` becomes multiple smaller kernels

**Review by rohan-varma — COMMENTED** [review]
Thanks for the PR! We've actually been working on some ideas for revamping optimizer overlapping which are in the design / discussion stage, and have prototyped many similar ideas as you've done here (discarding grads in the reducer is one example).

Regarding the initial traces you've provided above, they are super helpful, we will investigate why current overlapped interface is slower to launch kernels than the traditional method.

Wondering what your use case is and what sort of performance / memory wins you're seeing, and which real world workloads on your end may want to use optimizer overlapping? Thank you!

**Comment by li-yi-dong** [comment]
> Thanks for the PR! We've actually been working on some ideas for revamping optimizer overlapping which are in the design / discussion stage, and have prototyped many similar ideas as you've done here (discarding grads in the reducer is one example).
> 
> Regarding the initial traces you've provided above, they are super helpful, we will investigate why current overlapped interface is slower to launch kernels than the traditional method.
> 
> Wondering what your use case is and what sort of performance / memory wins you're seeing, and which real world workloads on your end may want to use optimizer overlapping? Thank you!

Hi @rohan-varma! 

As for 'why current overlapped interface is slower to launch kernels than the traditional method', I guess  it may due to the overhead for evaluating Python function in C++ via Pybind11 and the logic for launching kernels for Overlapped one is slightly more complicated than original one. 

For my use case, we are actually running the PyTorch1.8.2, of which would have a `stream synchronize` at the end of DDP backward, waiting for all computations on CUDA device then enter the Python frontend for Optimizer. Something like this:image
So that, by moving the kernel launching of Optimizer forward right after finishing `reduce bucket`, it can almost eliminate all the cost(in time) of Optimizer and have 5 to 10 percent gain. The actual benefits in terms of time consumption depends on:
1. How badly DDP bucketing the Grads and overlapping computation and communication.
2. How much computation the Optimizer and `zero_grad` needs (Adam is much heavy than SGD)
I didn’t pay attention to memory consumption 🙃

However, I notice that the latest version of PyTorch manage to eliminate the synchronization at the end of DDP backward, which enable the overlapping of Python frontend for Optimizer(typically consumes more time than the actual computation on device) and the `all_reduce`. I guess the improvement would drop to under 5 percent.

**Comment by li-yi-dong** [comment]
> Thanks for the PR! We've actually been working on some ideas for revamping optimizer overlapping which are in the design / discussion stage, and have prototyped many similar ideas as you've done here (discarding grads in the reducer is one example).
> 
> Regarding the initial traces you've provided above, they are super helpful, we will investigate why current overlapped interface is slower to launch kernels than the traditional method.
> 
> Wondering what your use case is and what sort of performance / memory wins you're seeing, and which real world workloads on your end may want to use optimizer overlapping? Thank you!

Can I join you gays for 'revamping optimizer overlapping'😃

**Comment by li-yi-dong** [comment]
@rohan-varma Hi, what do I need to merge this into master?

**Comment by kumpera** [comment]
@rohan-varma do you think we could merge this incrementally?

There are some part of this design that looks odd or bad to me, but other bits seem ready to be merged.

In particular, the DDP extension to discard gradients might be ready to be merged.

I don't like the overall design of overlapped optimizer as it requires reimplementing every optimizer to support overlapping, which is is quite limiting.

The changes to lr_scheduler and optimizer needs to be discussed separately as they are not clear from a quick glance.

**Comment by li-yi-dong** [comment]
> @rohan-varma do you think we could merge this incrementally?
> 
> There are some part of this design that looks odd or bad to me, but other bits seem ready to be merged.
> 
> In particular, the DDP extension to discard gradients might be ready to be merged.
> 
> I don't like the overall design of overlapped optimizer as it requires reimplementing every optimizer to support overlapping, which is is quite limiting.
> 
> The changes to lr_scheduler and optimizer needs to be discussed separately as they are not clear from a quick glance.

Hi @kumpera 

I agree that reimplementing every optimizer is a boilerplate, but the [this commit](https://github.com/pytorch/pytorch/commit/1c8fcc44cba0dcf38310637400e42bf6441c955c) also need to implement every `functional_optim`. 

I wrap `functional_optim` up and modify `lr_scheduler`, `optimizer` and `grad_scaler` to make them work together. To my knowledge, `lr_scheduler` is required for almost all training. I want to discuss the interface in [this feature request](https://github.com/pytorch/pytorch/issues/80595), but got no reply. So, I simply post this PR. I do think this PR need to be polished and appreciate any suggestion. 

Anyway, what I've done here is trying to make minimum modification on original `lr_scheduler` and `optimizer` meanwhile reuse as much existing code as I can.

**Comment by YLGH** [comment]
@li-yi-dong just curious, were you able to see memory reductions when releasing gradients early?

**Comment by li-yi-dong** [comment]
> @li-yi-dong just curious, were you able to see memory reductions when releasing gradients early?

I did not pay attention to that🙃
